src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs

```diff
@@ -0,0 +1,335 @@
+        /// </summary>
+        internal static readonly bool CanNotifyUpdates = !Utils.GetOptOutEnvVariableAsBool(UpdateCheckOptOutEnvVar, defaultValue: false)
+            && ExperimentalFeature.IsEnabled("PSUpdatesNotification");
+
+        // Maybe we shouldn't do update check and show notification when it's from a mini-shell, meaning when
+        // 'ConsoleShell.Start' is not called by 'ManagedEntrance.Start'.
+        // But it seems so unusual that it's probably not worth bothering. Also, a mini-shell probably should
+        // just disable the update notification feature by setting the opt-out environment variable.
+
+        internal static void ShowUpdateNotification(PSHostUserInterface hostUI)
+        {
+            if (!Directory.Exists(s_cacheDirectory))
+            {
+                return;
+            }
+
+            if (TryParseUpdateFile(
+                    updateFilePath: out _,
+                    out SemanticVersion lastUpdateVersion,
+                    lastUpdateDate: out _)
+               && lastUpdateVersion != null)
+            {
+                string releaseTag = lastUpdateVersion.ToString();
+                string notificationMsgTemplate = string.IsNullOrEmpty(lastUpdateVersion.PreReleaseLabel)
+                    ? ManagedEntranceStrings.StableUpdateNotificationMessage
+                    : ManagedEntranceStrings.PreviewUpdateNotificationMessage;
+
+                string notificationMsg = string.Format(CultureInfo.CurrentCulture, notificationMsgTemplate, releaseTag);
+                hostUI.WriteLine(notificationMsg);
+            }
+        }
+
+        internal static async Task CheckForUpdates()
+        {
+            // Delay the update check for 3 seconds so that it has the minimal impact on startup.
+            await Task.Delay(3000);
+
+            // A self-built pwsh for development purpose has the SHA1 commit hash baked in 'GitCommitId',
+            // which is 40 characters long. So we can quickly check the length of 'GitCommitId' to tell
+            // if this is a self-built pwsh, and skip the update check if so.
+            if (PSVersionInfo.GitCommitId.Length > 40)
+            {
+                return;
+            }
+
+            // If the host is not connect to a network, skip the rest of the check.
+            if (!NetworkInterface.GetIsNetworkAvailable())
+            {
+                return;
+            }
+
+            // Create the update cache directory if it doesn't exists
+            if (!Directory.Exists(s_cacheDirectory))
+            {
+                Directory.CreateDirectory(s_cacheDirectory);
+            }
+
+            bool parseSuccess = TryParseUpdateFile(
+                out string updateFilePath,
+                out SemanticVersion lastUpdateVersion,
+                out DateTime lastUpdateDate);
+
+            DateTime today = DateTime.UtcNow;
+            if (parseSuccess && updateFilePath != null && (today - lastUpdateDate).TotalDays < 7)
```

Comment:
A `TryXXXX` method can define its own criteria for success as well as the value for the out variables.
For APIs that are supposed to be reusable (public/internal), it's important to make them follow the pattern that is most familiar to people, just like what suggested here -- when `TryXXX` returns true, out values should be present.
But here, (1) it's to be used only here; (2) the main goal is to avoid any redundant code. In this case, even if we move out the existence check to a separate method, like `GetUpdateFile`, it still needs to return a bool along with a out parameter `updateFilePath` to differentiate `it's OK to not find any` and `it's not OK to find multiple` when `updateFilePath == null`.

If we are going to add the `postcondition` attribute, then it would `MaybeNullWhen` for this method.
```
        private static bool TryParseUpdateFile(
            [MaybeNullWhen(true)] out string? updateFilePath,
            [MaybeNullWhen(true)] out SemanticVersion? lastUpdateVersion,
            [MaybeNullWhen(true)] out DateTime? lastUpdateDate)
```

> (Ex., below we use if (!parseSuccess) but comment says "This could happen when there is no new updates yet." - this makes code difficult to understand)

I will improve the comments.